app/scripts/services/worker.js

```diff
@@ -321,14 +355,60 @@ function _uint16ArrayToFloat32Array(uint16array) {
 }
 
 /**
- * Convert a response from the tile server to
- * data that can be used by higlass
+ * @typedef TileData<Server>
+ * @property {string} server
+ * @property {string} tileId
+ * @property {number} zoomLevel
+ * @property {[number] | [number, number]} tilePos
+ * @property {string} tilesetUid
+ */
+
+/**
+ * @typedef DenseTileData
+ * @property {string} server
+ * @property {string} tileId
+ * @property {number} zoomLevel
+ * @property {[number] | [number, number]} tilePos
+ * @property {string} tilesetUid
+ * @property {Float32Array} dense
+ * @property {string} dtype
+ * @property {DenseDataExtrema1D | DenseDataExtrema2D} denseDataExtrema
+ * @property {number} minNonZero
+ * @property {number} maxNonZero
+ */
+
+/**
+ * @template T
+ * @typedef {Omit<T, keyof DenseTileData> & (TileData | DenseTileData)} CompletedTileData
  */
-export function tileResponseToData(data, server, theseTileIds) {
-  if (!data) {
-    // probably an error
-    data = {};
-  }
+
+/**
+ * @typedef TileResponse
+ * @property {string=} dense - a base64 encoded string
+ */
+
+/**
+ * Convert a response from the tile server to data that can be used by higlass.
+ *
+ * WARNING: Mutates the data object.
+ *
+ * @template {TileResponse} T
+ * @param {Record<string, T>} inputData
+ * @param {string} server
+ * @param {string[]} theseTileIds
+ *
+ * @returns {Record<string, CompletedTileData<T>>}
+ *
+ * Trevor: This function is littered with ts-expect-error comments because
+ * the type of mutation happening to the input object is very tricky to type.
+ * The type signature of the function tries to adequately describe the mutation,
+ * to outside users.
+ */
+export function tileResponseToData(inputData, server, theseTileIds) {
```

Comment:
NB, this function is really gnarly from a typing perspective. It's mutating the input objects in place. Easier to make copies, an override properties.

Reply:
An example implementation of this can be found in https://github.com/higlass/higlass/tree/manzt/safe-assign-tile-response-data